### ISSUE TITLE
How to avoid security vulnerabilities introduced in packages from compromised accounts

### ISSUE THREAD

**Issue by connorshea** [dependabot/feedback#425]
Given the bootstrap-sass Ruby gem backdoor that was introduced a week ago, I figure it'd be good to talk about ways this kind of attack can be mitigated: https://snyk.io/blog/malicious-remote-code-execution-backdoor-discovered-in-the-popular-bootstrap-sass-ruby-gem/

I'd like to propose a few possible things that would have prevented dependabot from updating anyone to this version:

- The version was only pushed to RubyGems, not to GitHub. Therefore, there was no mention of a 3.2.0.3 version in the Changelog or Releases/Tags. If Dependabot doesn't detect a Changelog or Release for that version, maybe it should ignore it, or at least not merge it automatically?
- Compare the SHA hash of the gem to the SHA hash of the actual git tag it corresponds to. Unfortunately this isn't very easy because Ruby gems frequently remove things from the repo (e.g. tests) when packaging the gem, and also because the SHA hash of the git commit is determined from the git history as well as the contents of the repository, so you'd need to 1) pull down the repo and get the SHA of its contents, and possibly 2) actually build the gem yourselves to get an actual comparable SHA hash for the gem. And there's not really a standard command that'd build every possible gem, and also this'd be quite compute-intensive.

Also, this all assumes the gem is open source, which is only the case >99% of the time. For closed source gems this wouldn't really work (though, does RubyGems actually host closed source gems themselves?)

Obviously, there's only so much that can be done on dependabot's side of things. If the attacker gains access to the maintainer's GitHub account as well, or if the maintainer were themselves malicious, these wouldn't actually mitigate the issue at all.

Thanks for reading, this isn't an easy problem but I wanted to bring it up :)

**Comment by greysteil** [comment]
Thanks for the discussion point @connorshea.

I'm going to come back to this, but I want to note immediately that no accounts using Dependabot were exposed to this vulnerability.

We wrote a post about the `event-stream` hijack issue a couple of months back [here](https://dependabot.com/blog/stopping-dependencies-getting-hijacked/). I'm going to revisit that with some more thoughts.

**Comment by greysteil** [comment]
Spent a bit more time thinking about this. I think building the gem and comparing the SHA of the built gem with the one published on Rubygems is the way to go. It's a major project on our side, but could be a meaningful contribution to security in the Ruby ecosystem.

**Motivation**
Having a canonical link between a release and its source code makes it *much* easier for users to check that the code is good. If even a few users are then diligent about checking the diff between dependency versions, malicious code should then be discovered much quicker. Dependabot already surfaces diffs. With the a way to check they match the compiled gem it would be even better, and there's a positive externality from the group of folks using Dependabot being diligent.

**Implementation thoughts**
Some unstructured thoughts on an option for implementing the above:
- Rubygems already exposes a checksum of the built gem (as pushed to it)
- Dependabot could therefore rebuild the gem based on the code in a tag (that it finds using its standard `MetadataFinder` logic), and then check whether it matches the sha in Rubygems. We'd be evaluating arbitrary code here, but we're already good at that
- We could cache the result on our side, so this shouldn't be too compute intensive. We might also be able to contribute it back to Rubygems somehow

**Comment by hmarr** [comment]
Yeah I’ve wanted to do this for a while, and built a proof of concept a couple months ago (which is where this is coming from) but we haven’t had time to get it finished yet. Fingers crossed we’ll be able to dedicate some time to it soon, though!

Fully reproducible build are tricky currently due to the way gems are built, particularly as the behaviour has changed in different RubyGems versions and we’d need to support old gems too. There are also come cases where the gem doesn’t include the gemspec on GitHub (Nokogiri, for instance).

All solvable problems though. @greysteil on the implementation, I think starting out with comparing what is present in the published gems, filtering for the known deterministic bits is a good place to start. This is what the initial version I built does.

Then working with the RubyGems folks to make builds properly reproducible would be a great direction to head in.

Ultimately, having something akin to Docker’s automated builds would be a great place to end up.

**Comment by eliotsykes** [comment]
This issue continues to be relevant.
